### ISSUE TITLE
Using GPU

### ISSUE THREAD

**Issue by clive-g-brown** [crystal-data/num.cr#24]
Having fun with this package, but cannot figure out how to move Tensors onto GPU instead of CPU. When I call tensor.opencl -> it stays on CPU.

yes I have a functioning clblas.

**Comment by christopherzimmerman** [comment]
Thanks for the issue.  Before, I was just grabbing the first available OpenCl device, which on my system is my GPU, but that isn't an optimal solution as seen here.  I've updated it to instead use the first GPU, with an added compile time flag `-Dopencl_any` if you want the first available device.

You will need to make sure the OpenCl shard updates, you might have to delete your shard lock / run shards update.

If that doesn't fix this, can you provide the output of `clinfo` on your machine?

**Comment by clive-g-brown** [comment]
thanks, ill have a go. yes it seems to default to CPU. I have two AMD GPUs like to get it to use.

**Comment by clive-g-brown** [comment]
Its only finding CPU, there are two OpenCL accessible GPUs [found by ArrayFire]. Would it be possible to have a function that enumerates the devices and lets you pick one by number ? Quite handy anyway.

unning
Pointer(Void)@0xffffffff
Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz

---------------------------------------------------------
VERSION = "0.1.0"

  # TODO: Put your code here

  puts "running"

  #a = [[1, 2], [3, 4]].to_tensor.astype(Float32)
  #b = [[3, 4], [5, 6]].to_tensor.astype(Float32)


  device, context, queue = Cl.single_device_defaults

  puts device
  puts Cl.device_name(device)


  #gpu = NumInternal::ClStorage(Float32).new(100000000)
  
  #z = ClTensor(Float32).new(gpu, [10000,10000])

  a = Tensor(Float32).zeros([10000,10000])
  b = Tensor(Float32).ones([10000,10000])

  acl = a.opencl
  bcl = b.opencl

  

  x = acl.matmul(bcl)

**Comment by christopherzimmerman** [comment]
Can you post the output of `clinfo`?

**Comment by clive-g-brown** [comment]
Its a Mac Pro - Catalina. 2X Vega64 II

Number of platforms                               1
  Platform Name                                   Apple
  Platform Vendor                                 Apple
  Platform Version                                OpenCL 1.2 (Feb 29 2020 00:40:07)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event

  Platform Name                                   Apple
Number of devices                                 3
  Device Name                                     Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz
  Device Vendor                                   Intel
  Device Vendor ID                                0xffffffff
  Device Version                                  OpenCL 1.2 
  Driver Version                                  1.1
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     CPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               32
  Max clock frequency                             3200MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             1024x1x1
  Max work group size                             1024
  Preferred work group size multiple              1
  Preferred / native vector sizes                 
    char                                                16 / 16      
    short                                                8 / 8       
    int                                                  4 / 4       
    long                                                 2 / 2       
    half                                                 0 / 0        (n/a)
    float                                                4 / 4       
    double                                               2 / 2        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              206158430208 (192GiB)
  Error Correction support                        No
  Max memory allocation                           51539607552 (48GiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        64
  Global Memory cache line size                   23068672 bytes
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            65536 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   1 bytes
    Pitch alignment for 2D image buffers          1 pixels
    Max 2D image size                             8192x8192 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Global
  Local memory size                               32768 (32KiB)
  Max number of constant args                     8
  Max constant buffer size                        65536 (64KiB)
  Max size of kernel argument                     4096 (4KiB)
  Queue properties                                
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      1ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            Yes
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels                                (n/a)
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_APPLE_fp64_basic_ops cl_APPLE_fixed_alpha_channel_orders cl_APPLE_biased_fixed_point_image_formats cl_APPLE_command_queue_priority

  Device Name                                     AMD Radeon Pro Vega II Compute Engine
  Device Vendor                                   AMD
  Device Vendor ID                                0x1021d00
  Device Version                                  OpenCL 1.2 
  Driver Version                                  1.2 (Mar  5 2020 22:38:16)
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               64
  Max clock frequency                             1731MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
  Preferred work group size multiple              64
  Preferred / native vector sizes                 
    char                                                 4 / 4       
    short                                                2 / 2       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 0 / 0        (n/a)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    32, Little-Endian
  Global memory size                              34342961152 (31.98GiB)
  Error Correction support                        No
  Max memory allocation                           4290772992 (3.996GiB)
  Unified memory for Host and Device              No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       32768 bits (4096 bytes)
  Global Memory cache type                        None
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   256 bytes
    Pitch alignment for 2D image buffers          256 pixels
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max number of constant args                     8
  Max constant buffer size                        65536 (64KiB)
  Max size of kernel argument                     1024
  Queue properties                                
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      40ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  printf() buffer size                            134217728 (128MiB)
  Built-in kernels                                (n/a)
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_image2d_from_buffer cl_khr_depth_images cl_APPLE_command_queue_priority cl_APPLE_command_queue_select_compute_units cl_khr_fp64

  Device Name                                     AMD Radeon Pro Vega II Compute Engine
  Device Vendor                                   AMD
  Device Vendor ID                                0x2021d00
  Device Version                                  OpenCL 1.2 
  Driver Version                                  1.2 (Mar  5 2020 22:38:16)
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               64
  Max clock frequency                             1731MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
  Preferred work group size multiple              64
  Preferred / native vector sizes                 
    char                                                 4 / 4       
    short                                                2 / 2       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 0 / 0        (n/a)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    32, Little-Endian
  Global memory size                              34342961152 (31.98GiB)
  Error Correction support                        No
  Max memory allocation                           4290772992 (3.996GiB)
  Unified memory for Host and Device              No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       32768 bits (4096 bytes)
  Global Memory cache type                        None
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   256 bytes
    Pitch alignment for 2D image buffers          256 pixels
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             2048x2048x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max number of constant args                     8
  Max constant buffer size                        65536 (64KiB)
  Max size of kernel argument                     1024
  Queue properties                                
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      40ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
  printf() buffer size                            134217728 (128MiB)
  Built-in kernels                                (n/a)
  Device Extensions                               cl_APPLE_SetMemObjectDestructor cl_APPLE_ContextLoggingFunctions cl_APPLE_clut cl_APPLE_query_kernel_names cl_APPLE_gl_sharing cl_khr_gl_event cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_image2d_from_buffer cl_khr_depth_images cl_APPLE_command_queue_priority cl_APPLE_command_queue_select_compute_units cl_khr_fp64

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Apple
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [P0]
  clCreateContext(NULL, ...) [default]            Success [P0]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (2)
    Platform Name                                 Apple
    Device Name                                   AMD Radeon Pro Vega II Compute Engine
    Device Name                                   AMD Radeon Pro Vega II Compute Engine
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  Success (1)
    Platform Name                                 Apple
    Device Name                                   Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (2)
    Platform Name                                 Apple
    Device Name                                   AMD Radeon Pro Vega II Compute Engine
    Device Name                                   AMD Radeon Pro Vega II Compute Engine
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  Invalid device type for platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (3)
    Platform Name                                 Apple
    Device Name                                   AMD Radeon Pro Vega II Compute Engine
    Device Name                                   AMD Radeon Pro Vega II Compute Engine
    Device Name                                   Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz
	NOTE:	your OpenCL library only supports OpenCL 1.0,
		but some installed platforms support OpenCL 1.2.
		Programs using 1.2 features may crash
		or behave unexpectedly

**Comment by christopherzimmerman** [comment]
I believe that this is the issue:

```
NOTE: your OpenCL library only supports OpenCL 1.0,
but some installed platforms support OpenCL 1.2.
Programs using 1.2 features may crash
or behave unexpectedly
```

You will need an opencl library that supports 1.2.  To verify this is the case, can you run the following snippet:

```
require "opencl"
platform = Cl.first_platform
devices = Cl.get_devices(platform)
devices.each do |d|
  puts Cl.device_name(d)
end
```

**Comment by clive-g-brown** [comment]
Thanks - here's the result [BTW its fine with ArrayFire]

Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz
AMD Radeon Pro Vega II Compute Engine
AMD Radeon Pro Vega II Compute Engine

**Comment by christopherzimmerman** [comment]
Ah, I think I might know.  You are calling `single_device_defaults` in your snippet, which isn't actually being used as the device.

Try:

```
require "opencl"
puts Cl.device_name(Num::ClContext.instance.device)
```

^ That is the device actually being used.

**Comment by clive-g-brown** [comment]
OK, I ran it without that [ that's from when I was trying to figure out why its only choosing the CPU]

**Comment by christopherzimmerman** [comment]
What is the exact command you are using to run the crystal script? And I am adding in a way to manually select the device now.

**Comment by clive-g-brown** [comment]
I do shards build - then run the binary.

  VERSION = "0.1.0"

  # TODO: Put your code here

  puts "running"

 puts Cl.device_name(Num::ClContext.instance.device)
  
 #platform = Cl.first_platform
 #devices = Cl.get_devices(platform)
 #devices.each do |d|
 #  puts Cl.device_name(d)
 # end

  #a = [[1, 2], [3, 4]].to_tensor.astype(Float32)
  #b = [[3, 4], [5, 6]].to_tensor.astype(Float32)



  #gpu = NumInternal::ClStorage(Float32).new(100000000)
  
  #z = ClTensor(Float32).new(gpu, [10000,10000])

  a = Tensor(Float32).zeros([10000,10000])
  b = Tensor(Float32).ones([10000,10000])

  acl = a.opencl
  bcl = b.opencl

  

  x = acl.matmul(bcl)

**Comment by christopherzimmerman** [comment]
Are you using the `-Dopencl_any` flag? If so, remove it, that will pick the CPU

**Comment by clive-g-brown** [comment]
shards build -Dopencl_any

still gives : 

running
Intel(R) Xeon(R) W-3245 CPU @ 3.20GHz

**Comment by christopherzimmerman** [comment]
You should be able to set a device now.  Try the following snippet:

```
require "./num"
require "opencl"
platform = Cl.first_platform
devices = Cl.get_devices(platform)
Num::ClContext.set_device(devices[1])  # or replace with whichever device you want.
puts Cl.device_name(Num::ClContext.instance.device)
```

**Comment by clive-g-brown** [comment]
thanks, I rm shard.lock, then shards update and shards build

Error target neon failed to compile:
Showing last frame. Use --error-trace for full trace.

There was a problem expanding macro 'class_getter'

Code in macro 'macro_4656982992'

 2 | class_getter instance : NumInternal::ClInfo { NumInternal::ClInfo.new(*Cl.first_gpu_defaults) }
     ^
Called macro defined in macro 'macro_4607268272'

 1052 | macro class_getter(*names, &block)

Which expanded to:

 >  9 | def self.instance
 > 10 |   if (value = @@instance).nil?
 > 11 |     @@instance = begin NumInternal::ClInfo.new(*Cl.first_gpu_defaults) end
                                                           ^-----------------
Error: undefined method 'first_gpu_defaults' for Cl:Module

**Comment by clive-g-brown** [comment]
Ta Da !!running
AMD Radeon Pro Vega II Compute Engine

**Comment by christopherzimmerman** [comment]
Awesome! I appreciate your patience.  If I'm correct, you shouldn't even need to manually set a device anymore, just the default should now pick up the GPU.

**Comment by clive-g-brown** [comment]
Yes it does - very nice and -Dopencl_any goes back to CPU

**Comment by clive-g-brown** [comment]
having a device picker is useful.

**Comment by christopherzimmerman** [comment]
Closed by #26
